### PR TITLE
Add support for building subset of images

### LINKED ISSUE #24: Add support for building a subset of the manifest.
This is related to https://github.com/aspnet/aspnet-docker/issues/259.  This functionality is need to support breaking apart the Windows builds into separate legs in order to decrease the overall build times.

### LINKED ISSUE #25: Image-builder logic to explicitly pull FROM images during build needs to support multi-stage builds
@MichaelSimons commented on [Thu Jul 06 2017](https://github.com/dotnet/dotnet-docker-nightly/issues/345)

The logic within image-builder that explicitly pulls the FROM images during a build needs to be updated to support multi-stage builds.  Multi-stage build Dockerfiles may have multiple FROM commands.  image-builder should be explicitly pulling each one of the base images.

### PR DESCRIPTION
**Pull request by MichaelSimons**
- Add support to build Dockerfiles for a specific repo
- Add support to build Dockerfiles within a specific path
- Refactored logic that detects FROM image to support the mult-stage builds feature which allows multiple FROM commands within a Dockerfile

Related #24 
Related #25

### CODE REVIEW

File: src/Microsoft.DotNet.ImageBuilder/ViewModel/PlatformInfo.cs [hunk 1]
```diff
@@ -39,13 +38,14 @@ public static PlatformInfo Create(Platform model, Manifest manifest, string repo
 
         private void InitializeFromImage()
         {
-            Match fromMatch = FromRegex.Match(File.ReadAllText(Path.Combine(Model.Dockerfile, "Dockerfile")));
-            if (!fromMatch.Success)
+            IEnumerable<Match> fromMatches = FromRegex.Matches(File.ReadAllText(Path.Combine(Model.Dockerfile, "Dockerfile")))
```
Comment (dagood) [inline src/Microsoft.DotNet.ImageBuilder/ViewModel/PlatformInfo.cs]: \<nit> Long line

File: src/Microsoft.DotNet.ImageBuilder/ImageBuilder.cs [hunk 2]
```diff
@@ -74,10 +74,13 @@ private static void BuildImages()
             foreach (ImageInfo image in Manifest.Images.Where(image => image.Platform != null))
             {
                 Console.WriteLine($"-- BUILDING: {image.Platform.Model.Dockerfile}");
-                if (!Options.IsSkipPullingEnabled && image.Platform.IsExternalFromImage)
+                if (!Options.IsSkipPullingEnabled)
                 {
-                    // Ensure latest base image exists locally before building
-                    ExecuteHelper.ExecuteWithRetry("docker", $"pull {image.Platform.FromImage}", Options.IsDryRun);
+                    // Ensure latest base images exist locally before building
+                    foreach (string fromImage in image.Platform.FromImages.Where(from => Manifest.IsExternalImage(from)))
```
Comment (dagood) [inline src/Microsoft.DotNet.ImageBuilder/ImageBuilder.cs]: Potentially use `Manifest.IsExternalImage` as a method group.

File: src/Microsoft.DotNet.ImageBuilder/ViewModel/PlatformInfo.cs (hunk 1 above)
Comment (MichaelSimons) [inline src/Microsoft.DotNet.ImageBuilder/ViewModel/PlatformInfo.cs]: Split

File: src/Microsoft.DotNet.ImageBuilder/ImageBuilder.cs (hunk 2 above)
Comment (MichaelSimons) [inline src/Microsoft.DotNet.ImageBuilder/ImageBuilder.cs]: Changed

**Comment by MichaelSimons** [comment]
Testing with dotnet-docker-nightly revealed the need to be able to pass variables into the test commands.  When building a subset of the Dockerfiles there needs to be a way to filter which tests are run.

`dotnet Microsoft.DotNet.ImageBuilder.dll --command build --path 2.0 --test-var TestFilter=2.0`

@dagood - can you please review the latest changes?

File: src/Microsoft.DotNet.ImageBuilder/Options.cs [hunk 3]
```diff
@@ -97,16 +119,50 @@ private static string GetArgValue(string[] args, ref int i, string argName)
...
+
         private static bool IsNextArgValue(string[] args, int i)
         {
             return i + 1 < args.Length && !args[i + 1].StartsWith("-");
         }
+
+        private static IDictionary<string, string> ParseNameValuePairs(IEnumerable<string> nameValuePairs)
+        {
+            Dictionary<string, string> result = new Dictionary<string, string>();
+            foreach (string nameValuePair in nameValuePairs) {
```
Comment (dagood) [inline src/Microsoft.DotNet.ImageBuilder/Options.cs]: Brace on same line.

File: src/Microsoft.DotNet.ImageBuilder/Options.cs [hunk 4]
```diff
@@ -97,16 +119,50 @@ private static string GetArgValue(string[] args, ref int i, string argName)
...
+                string[] splitResults = nameValuePair.Split(new char[] { '=' }, 2);
+                result.Add(splitResults[0], splitResults[1]);
+            }
+
+            return result;
```
Comment (dagood) [inline src/Microsoft.DotNet.ImageBuilder/Options.cs]: Maybe:
```C#
return nameValuePairs
    .Select(pair => pair.Split(new char[] { '=' }, 2))
    .ToDictionary(split => split[0], split => split[1])
```

File: src/Microsoft.DotNet.ImageBuilder/Options.cs (hunk 3 above)
Comment (MichaelSimons) [inline src/Microsoft.DotNet.ImageBuilder/Options.cs]: fixed

File: src/Microsoft.DotNet.ImageBuilder/Options.cs (hunk 4 above)
Comment (MichaelSimons) [inline src/Microsoft.DotNet.ImageBuilder/Options.cs]: good suggestion - changed

**Pull request merged**